acinclude/printf.m4

```diff
@@ -0,0 +1,30 @@
+## Copyright (C) 1996-2023 The Squid Software Foundation and contributors
+##
+## Squid software is distributed under GPLv2+ license and includes
+## contributions from numerous individuals and organizations.
+## Please see the COPYING and CONTRIBUTORS files for details.
+##
+
+dnl check which printf code to use for a type
+dnl $1 - the type being checked
+dnl $2 - the PRIxxx macro to define
+dnl $3 - a list of possible codes
+AC_DEFUN([SQUID_CHECK_PRINTF_CODES],[
+  AH_TEMPLATE([$2],[printf display of $1])
+  AC_CACHE_CHECK([which printf code displays $1],
+                 [squid_cv_printf_$2],[
+    SQUID_STATE_SAVE($2)
+    CXXFLAGS="-Werror=format"
```

Comment:
We cannot expect all compilers to support that option. I am worried that when the compiler ignores this option, we may build Squid with the wrong printf `%code`, and when the compiler refuses to run with that option, we may not find the right printf() code at all.

Also, we should not just completely overwrite CXXFLAGS because that may break building the test program below when building _anything_ requires custom CXXFLAGS (supplied by the admin).

Since the first given %code is supposed to be the most portable/universal, and since `-Wformat` is enabled by `-Wall` in GCC, I suggest removing this CXXFLAGS customization. We may still use the wrong %code in some cases, but we will cover more cases with simpler code this way. Moreover, the official code is already using this approach AFAICT.

```suggestion
```

If removing this line means that we can also remove SQUID_STATE_SAVE/SQUID_STATE_ROLLBACK, please remove those as well, of course.

Reply:
Regarding compiler support: All the compilers we officially support accept this flag. If we need to add support for one that has different options we can adjust this line at that time.

Why we need it: this format check is disabled in the current set of flags used Squid-wide. Without it any of the options that size_t can implicitly convert to (regardless of value wrap/loss) will be detected as valid format codes.

I will fix this to be an append to user flags though. Thanks for that catch.

Reply:
> All the compilers we officially support accept this flag.

I do not recall us agreeing on the list of supported compilers but, more importantly, the primary problem here is _not_ with the compilers we test with:

> If we need to add support for one that has different options we can adjust this line at that time.

The problem here is not so much with adjusting when we _know_ that we need to adjust. The problem is an admin building Squid with a compiler that does not support this option, with the build _silently_ failing to detect the right %code and using the wrong %code because of that lack of support (without our or admin knowledge about any of that).

> this format check is disabled in the current set of flags used Squid-wide

What makes you think that? GCC documentation says that `-Wall` enables `-Wformat`. Clang [documentation](https://clang.llvm.org/docs/DiagnosticsReference.html#wformat) says that `-Wformat` is enabled by default, and my [test](https://godbolt.org/z/r5YPd3sn6) supports that claim. Squid enables `-Wall` and `-Werror` by default (where supported).

> Without it any of the options that size_t can implicitly convert to (regardless of value wrap/loss) will be detected as valid format codes.

I do not think that is an accurate assumption because of the -Wall and defaults discussed above.

In summary, explicitly adding `-Wformat` is not needed for common Squid builds and may create/cause problems in some environments. Omitting that option may create problems in some other environments, but those problems are likely to extend to the whole Squid code, not just this test case. Thus, it looks like we should not add this compiler flag here. When/if we discover specific environments where it (or its equivalent) is needed, we will enable it globally for those cases, not just here.

However, perhaps I am missing something that has not been (dis)covered yet. If you remove this configure.ac line, does the overall outcome change in your MinGW test environment? If it does, there may be another bug we need to fix somewhere!

Reply:
Done in [3d746d2](https://github.com/squid-cache/squid/pull/1387/commits/3d746d2cdce91fbae23109ef3faaf1b5c2803c61).

FTR; I am going by https://github.com/squid-cache/squid/blob/master/doc/release-notes/release-6.sgml.in#L27 for my definition of "officially supported compilers".

This line was tested with GCC 12 in both MinGW-w64 cross-build and Linux native build. Using default ```./configure``` options. Several order permutations of the format selection and also ```%az``` (invalid) were used. Only the explicit addition of this flag caused ``%u``` to produce the expected error about bad down-casting on my 64-bit machine and continuing through to find ```%zu``` or ```%llu``` whichever was next (those being equivalent on 64-bit builds).
 There are some weird things going on with flags set (or not) by some tests, and misuse of some flags in our configure scripts. For now I think this is "good enough" and removal can wait on future work cleaning up that mess.

Reply:
Moved to https://github.com/squid-cache/squid/pull/1387#discussion_r1237259277